<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

Прочитаем файл:

In [2]:
df = pd.read_csv('/datasets/insurance.csv')                

Посмотрим его размер и пропущенные значения:

In [3]:
print(df.shape)                                             
df.isnull().sum()                                            

(5000, 5)


Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Посмотрим сам файл:

In [4]:
display(df.head(5))                                           

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Посмотрим есть ли дубликаты:

In [5]:
df.duplicated().sum()                                     

153

Удаляем обнаруженные дубликаты

In [6]:
df = df.drop_duplicates().reset_index(drop=True)

### Вывод

На первом этапе было сделано:
- открыт и изучен файл, в нем содержиться 5000 строк и 5 столбцов,
- проверили на наличие пропущенных значений и дубликатов, удалили 153 дубликата.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** $ R^2 $ не измениться
      
    
**Обоснование:**   
$$w'  = ((XP)^T (XP))^{-1} (XP)^T y  =  (P^T X^T X P)^{-1}  (XP)^T y  =  (P^T (X^TX) P)^{-1} (XP)^T y  =  (P^T ((X^TX) P))^{-1} P^T X^T y  =  ((X^TX)P)^{-1} (((P^T))^{-1}P^T) X^T y  =  ((X^TX)P)^{-1} E X^T y  =  P^{-1}(X^TX)^{-1} X^Ty$$

$$w' = P^{-1}(X^TX)^{-1} X^Ty$$
$$w' = P^{-1}w$$

## Алгоритм преобразования

**Алгоритм**

1. Создать обратную рандомную матрицу применив numpy.random.normal() (при этом вероятность вероятность получить необратимую матрицу близка к нулю, но проверим)
2. Перемножим матрицы - то есть проведем шифрование
3. Проведем дешифрование и проверим соответствие первых 5 строк
4. Проверим качество модели машинного обучения

**Обоснование**

После выполнения алгоритма без наличия нашей рандомной матрицы восстановить персональную информацию клиентов страховой компании будет проблематично. Качество модели не должно ухудшиться.

## Проверка алгоритма

In [7]:
features = df.drop('Страховые выплаты', axis = 1)
target = df['Страховые выплаты']

Посмотрим размеры выборок

In [8]:
features.shape, target.shape

((4847, 4), (4847,))

In [9]:
features_train, features_valid, target_train, target_valid = train_test_split(features, 
                                                                              target, test_size=0.25, random_state=12345)

Создадим и посмотрим матрицу:

In [10]:
matrix = np.random.normal(0, 157, size = (4,4))                

In [11]:
print(matrix)                                                   

[[  31.60983465  -65.15424958  129.6839464     9.06599883]
 [-505.37664261  141.90919565  -57.67398017    8.36787265]
 [-197.96247707 -175.70546028   57.57882937 -127.28132757]
 [ 283.16537952 -184.72033738  -64.70203104   41.01135727]]


Проверим ее на обратимость, ошибки не получено, значит все OK

In [12]:
back_matrix = np.linalg.inv(matrix)                           
print(back_matrix)

[[-0.00123879 -0.0023362  -0.00051996 -0.00086322]
 [-0.002328   -0.00196409 -0.00173049 -0.00445531]
 [ 0.00628527 -0.0008345  -0.00040418 -0.00247357]
 [ 0.00798369  0.00596734 -0.00484189  0.00637393]]


In [13]:
def to_r2_score(features, target):
    lr = LinearRegression()
    lr.fit(features, target)
    r2 = lr.score(features, target)
    print(r2)

In [14]:
print("Посчитаем R2 для исходной матрицы:")
to_r2_score(features, target)

Посчитаем R2 для исходной матрицы:
0.4302010044852067


In [15]:
df_encrypted = features @ matrix
print("Посчитаем R2 для матрицы, полученной после перемножения:")
to_r2_score(df_encrypted, target)

Посчитаем R2 для матрицы, полученной после перемножения:
0.4302010044852016


In [16]:
print(df_encrypted)

                 0             1             2             3
0    -9.839345e+06 -8.709422e+06  2.853610e+06 -6.312761e+06
1    -7.545538e+06 -6.670464e+06  2.185278e+06 -4.836265e+06
2    -4.171868e+06 -3.685699e+06  1.207483e+06 -2.672665e+06
3    -8.265082e+06 -7.324307e+06  2.399697e+06 -5.307374e+06
4    -5.180940e+06 -4.582004e+06  1.501322e+06 -3.321799e+06
...            ...           ...           ...           ...
4842 -7.080845e+06 -6.269081e+06  2.053820e+06 -4.543627e+06
4843 -1.039013e+07 -9.202326e+06  3.015105e+06 -6.669216e+06
4844 -6.720469e+06 -5.953946e+06  1.950639e+06 -4.314588e+06
4845 -6.483610e+06 -5.743066e+06  1.881494e+06 -4.161783e+06
4846 -8.051112e+06 -7.129918e+06  2.336151e+06 -5.167338e+06

[4847 rows x 4 columns]


По полученной таблице не понятно какие действительные значения должны быть.

Посмотрим сможем ли получить обратно исходную матрицу иначе смысла в действиях нет и сравним ее с исходной.

In [17]:
back = df_encrypted @ back_matrix
display(np.around(back, 0).head(5))

0     1        2    3
0  1.0  41.0  49600.0  1.0
1 -0.0  46.0  38000.0  1.0
2  0.0  29.0  21000.0  0.0
3  0.0  21.0  41700.0  2.0
4  1.0  28.0  26100.0 -0.0

In [18]:
display(df.head(5))                                            # исходная таблица

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Сравнив таблицы, видим что они одинаковые.

### Вывод

Защитили данные клиентов страховой компании, для этого произвели умножение исходных данных на рандомную матрицу. Перед этим проверили рандомную матрицу на обратимость. После произвели обратное преобразование в исходную матрицу для проверки, что с данными можно продолжить работать. 

Проверили качество преобразования по R2, качество исходной матрицы и полученной после умножения одинаковое. 

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования